presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizer.java

```diff
@@ -75,27 +83,41 @@ public void shutdown()
         }
     }
 
-    public CompletableFuture<?> enqueue(OrganizationSet organizationSet)
+    public ListenableFuture<?> enqueue(OrganizationSet organizationSet)
     {
+        log.info("enqueue organizationSet: %s", organizationSet);
         shardsInProgress.putAll(organizationSet.getShardsMap());
-        return runAsync(jobFactory.create(organizationSet), executorService)
-                .whenComplete((none, throwable) -> {
+        ListenableFuture<?> listenableFuture = prioritizedFifoExecutor.submit(jobFactory.create(organizationSet));
+        listenableFuture.addListener(
+                () -> {
```

Comment:
Instead of adding 2 levels of callback, can you extract the for loop into a function and call it at `onSuccess()` and `onFailure()` below?

Reply:
But that would add duplicate code, Let's also confirm with James before we make this change

Reply:
Why would this create duplicate code?